src/libraries/System.Console/src/System/IO/KeyParser.cs

```diff
@@ -0,0 +1,387 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.IO;
+
+internal static class KeyParser
+{
+    private const char Escape = '\u001B';
+    private const char Delete = '\u007F';
+    private const char VtSequenceEndTag = '~';
+    private const char ModifierSeparator = ';';
+    private const int MinimalSequenceLength = 3;
+    private const int SequencePrefixLength = 2; // ^[[ ("^[" stands for Escape)
+
+    internal static ConsoleKeyInfo Parse(char[] buffer, TerminalFormatStrings terminalFormatStrings, byte posixDisableValue, byte veraseCharacter, ref int startIndex, int endIndex)
+    {
+        int length = endIndex - startIndex;
+        Debug.Assert(length > 0);
+
+        // VERASE overrides anything from Terminfo. Both settings can be different for Linux and macOS.
+        if (buffer[startIndex] != posixDisableValue && buffer[startIndex] == veraseCharacter)
+        {
+            // the original char is preserved on purpose (backward compat + consistency)
+            return new ConsoleKeyInfo(buffer[startIndex++], ConsoleKey.Backspace, false, false, false);
+        }
+
+        // Escape Sequences start with Escape. But some terminals like PuTTY and rxvt use Escape to express that for given sequence Alt was pressed.
+        if (length >= MinimalSequenceLength + 1 && buffer[startIndex] == Escape && buffer[startIndex + 1] == Escape)
+        {
```

Comment:
It could be, if it would be possible to identify PuTTy settings on our side.

With PuTTy, the value of `TERM` env var is configured with the following setting:

![image](https://user-images.githubusercontent.com/6011991/182133264-668cdf86-3791-4c8c-9165-61975376db12.png)

which has the default value of `xterm` (so on our side we think the user is using `xterm` terminal), but can be configured to `putty`, `linux` or any other know term.

On top of that, users can request PuTTy to use different key mappings. For example Linux text console or rxvt:

![image](https://user-images.githubusercontent.com/6011991/182133710-ae1131e1-3038-4299-823c-6589a455713d.png)

And PuTTy does not set any env vars to let us know that it's using some custom settings.